feat(MA): add conversion based on last touching-point

jabahamondes opened the pull request:
## Problem

Giving attribution to conversion goal that don't contain utm parameters.

## Changes

For this, the approach was to, by default (in the future configurable):
- get the conversion
- search for $pageviews events with utm parameters previous the conversion based on a 365 days windows (for now)
- attribute the conversion to the utm parameters from the last (for now) pageview

## How did you test this code?

Manually and added more tests

Example of final query:

```

  SELECT
      if(notEmpty(campaign_name), campaign_name, 'organic') AS campaign,
      if(notEmpty(source_name), source_name, 'organic') AS source,
      count() AS conversion_0
  
  FROM
      (
  SELECT
          person_id,
          if(notEmpty(conversion_campaign), conversion_campaign, if(notEmpty(fallback_campaign), fallback_campaign, '')) AS campaign_name,
          if(notEmpty(conversion_source), conversion_source, if(notEmpty(fallback_source), fallback_source, '')) AS source_name,
          1 AS conversion_value
      
  FROM
          (
  SELECT
              person_id,
              conversion_timestamps[i] AS conversion_time,
              conversion_math_values[i] AS conversion_math_value,
              conversion_campaigns[i] AS conversion_campaign,
              conversion_sources[i] AS conversion_source,
              arrayMax(arrayFilter(x -> and(lessOrEquals(x, conversion_timestamps[i]), greaterOrEquals(x, minus(conversion_timestamps[i], 15552000))), utm_timestamps)) AS last_utm_timestamp,
              if(notEquals(last_utm_timestamp, NULL), utm_campaigns[indexOf(utm_timestamps, last_utm_timestamp)], '') AS fallback_campaign,
              if(notEquals(last_utm_timestamp, NULL), utm_sources[indexOf(utm_timestamps, last_utm_timestamp)], '') AS fallback_source
          
  FROM
              (
  SELECT
                  events.person_id,
                  arrayFilter(x -> greater(x, 0), groupArray(if(equals(events.event, 'user signed up'), toUnixTimestamp(events.timestamp), 0))) AS conversion_timestamps,
                  arrayFilter(x -> greater(x, 0), groupArray(if(equals(events.event, 'user signed up'), toFloat(1), 0))) AS conversion_math_values,
                  arrayFilter(x -> notEmpty(toString(x)), groupArray(if(equals(events.event, 'user signed up'), toString(ifNull(events.properties.utm_campaign, '')), ''))) AS conversion_campaigns,
                  arrayFilter(x -> notEmpty(toString(x)), groupArray(if(equals(events.event, 'user signed up'), toString(ifNull(events.properties.utm_source, '')), ''))) AS conversion_sources,
                  arrayFilter(x -> greater(x, 0), groupArray(if(and(equals(events.event, '$pageview'), notEmpty(toString(ifNull(events.properties.utm_campaign, ''))), notEmpty(toString(ifNull(events.properties.utm_source, '')))), toUnixTimestamp(events.timestamp), 0))) AS utm_timestamps,
                  arrayFilter(x -> notEmpty(x), groupArray(if(and(equals(events.event, '$pageview'), notEmpty(toString(ifNull(events.properties.utm_campaign, ''))), notEmpty(toString(ifNull(events.properties.utm_source, '')))), toString(ifNull(events.properties.utm_campaign, '')), ''))) AS utm_campaigns,
                  arrayFilter(x -> notEmpty(x), groupArray(if(and(equals(events.event, '$pageview'), notEmpty(toString(ifNull(events.properties.utm_campaign, ''))), notEmpty(toString(ifNull(events.properties.utm_source, '')))), toString(ifNull(events.properties.utm_source, '')), ''))) AS utm_sources
              
  FROM
                  events
              
  WHERE
                  or(and(equals(events.event, 'user signed up'), greaterOrEquals(events.timestamp, toDateTime('2025-06-06')), lessOrEquals(events.timestamp, toDateTime('2025-07-07'))), and(equals(events.event, '$pageview'), notEmpty(toString(ifNull(events.properties.utm_campaign, ''))), notEmpty(toString(ifNull(events.properties.utm_source, ''))), greaterOrEquals(events.timestamp, minus(toDateTime('2025-06-06'), toIntervalSecond(15552000))), lessOrEquals(events.timestamp, toDateTime('2025-07-07'))))
              
  GROUP BY
                  events.person_id
              
  HAVING
                  greater(length(conversion_timestamps), 0))
          ARRAY JOIN arrayEnumerate(conversion_timestamps) AS i)) AS attributed_conversions
  
  GROUP BY
      campaign,
      source
  
  LIMIT 100
  ```

Commit:
feat: Refactor ConversionGoalFilter schema and related components

- Removed deprecated ActionFilter schema from schema.json.
- Updated ConversionGoalFilter to support new properties and structure, including event handling and fixed properties.
- Adjusted ConversionGoalDropdown and ConversionGoalsConfiguration components to align with the new filter structure.
- Enhanced validation logic for conversion goals in the backend to accommodate new NodeKind types.
- Cleaned up unused imports and code in various files for better maintainability.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

jabahamondes committed:
Working first basic (#34116)

Co-authored-by: Javier Bahamondes <javierbahamondes@Javiers-MacBook-Pro.local>

Commit:
Merge conflicts fix

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

Commit:
Prs comments

Commit:
Merge branch 'feat/conversion-goal-v2-ma' of github.com:PostHog/posthog into feat/conversion-goal-v2-ma

github-actions[bot] committed:
Update UI snapshots for `chromium` (13)

github-actions[bot] committed:
Update query snapshots

Commit:
Sanitize fn in utils

Commit:
Refactor marketing analytics query generation to use AST internally

- Updated , , , and other methods to utilize AST for building SQL queries.
- Introduced new methods for generating campaign cost CTE and base columns using AST.
- Adapted marketing source adapters (BigQuery, Google Ads, Self-Managed) to return AST expressions for field retrieval.
- Enhanced date range condition handling with AST for better query construction.

This change improves the maintainability and readability of the query generation logic.

Commit:
More Hogql migration code

Commit:
Add MarketingAnalyticsFilters component and integrate into WebAnalyticsDashboard

- Introduced MarketingAnalyticsFilters component for marketing-specific date filtering.
- Updated WebAnalyticsDashboard to conditionally render MarketingAnalyticsFilters based on the selected product tab.
- Modified WebAnalyticsHealthCheck to return null for the marketing tab, preventing unnecessary checks.
- Refactored conversion goal processor to enhance query generation with AST, improving maintainability.

This update enhances the user experience for marketing analytics by providing tailored filtering options.

Commit:
Refactor marketing analytics query generation and improve AST usage

- Updated  to streamline UTM expression retrieval and condition handling.
- Simplified query generation methods to return AST directly, enhancing maintainability.
- Refactored marketing source adapters to align with new method signatures for field retrieval.
- Improved date range condition handling across adapters for better query construction.

This update enhances the overall structure and readability of the marketing analytics query generation logic.

Commit:
Merge branch 'feat/conversion-goal-v2-ma' of github.com:PostHog/posthog into feat/conversion-goal-v2-ma

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

Commit:
Fix formatting issues in MarketingAnalyticsFilters and WebAnalyticsDashboard; refactor conversion goal processor method names for clarity

- Added missing newline at the end of MarketingAnalyticsFilters.tsx.
- Corrected the closing brace in WebAnalyticsDashboard.tsx for proper syntax.
- Renamed method in ConversionGoalProcessor from  to  to better reflect its functionality and updated related calls.

These changes improve code readability and maintainability across the marketing analytics components.

Commit:
Merge remote-tracking branch 'origin/master' into feat/conversion-goal-v2-ma

Commit:
Merge branch 'origin/feat/conversion-goal-v2-ma' into feat/conversion-goal-v2-ma

Commit:
Fixing bot comments

Commit:
Fixing all sort of types

Commit:
Fixing conflicts

Commit:
migrating all to ast expressions

Commit:
Merge remote-tracking branch 'origin/master' into feat/conversion-goal-v2-ma

Commit:
Creating conversion goal filter

Commit:
style improvements and moving the filter into the table

Commit:
Merge remote-tracking branch 'origin/master' into feat/conversion-goal-v2-ma

Commit:
Merge branch 'feat/conversion-goal-v2-ma' into feat/creating-conversion-goal-filter-in-dashboard

Commit:
fixing merge conflicts

Commit:
refactor: Update MarketingAnalyticsTable to integrate ConversionGoalFilter and enhance goal handling

- Removed the old MarketingAnalyticsTable component and replaced it with an updated version that includes ConversionGoalFilter.
- Improved handling of conversion goals by ensuring both static and dynamic goals are combined correctly.
- Cleaned up code by removing unnecessary comments and ensuring type compatibility for conversion goals.

This change aims to streamline the marketing analytics functionality and improve overall performance.

Commit:
Fixing CI

github-actions[bot] committed:
Update query snapshots

Commit:
feat: Enhance DynamicConversionGoalControls with expandable UI for conversion goals

- Added a button to toggle the visibility of the conversion goal dropdown.
- Integrated icons for better UX feedback on the button state.
- Refactored the layout to conditionally render the dropdown and action buttons based on the expanded state.

This update improves user interaction by allowing users to expand and collapse the conversion goal options, streamlining the interface.

Commit:
functional refactor

Commit:
Conversion goal enums refactor

Commit:
refactor: Update Marketing Analytics schema and types

- Renamed and reorganized Marketing Analytics enums and schemas for clarity and consistency.
- Introduced new types for MarketingAnalyticsSchemaField and MarketingAnalyticsSchemaFieldTypes.
- Updated various components and logic to utilize the new schema structure, enhancing type safety and maintainability.
- Removed deprecated fields and streamlined the source mapping process.

This refactor aims to improve the overall structure and usability of the marketing analytics components.

Commit:
refactor: Update source mapping in Marketing Analytics

- Introduced  property to convert raw source mappings into  objects for better type handling.
- Updated  and  methods to accept  instead of  for improved flexibility.
- Removed the  utility function in favor of direct attribute access for clarity and performance.
- Adjusted various adapters to utilize the new source mapping structure, enhancing maintainability and type safety.

This refactor aims to streamline the source mapping process and improve the overall structure of the marketing analytics components.

Commit:
fixing eslint'

Commit:
feat: Enhance DynamicConversionGoalControls with new input and save functionality

- Added a text input for custom conversion goal names to improve user experience.
- Implemented save functionality to persist new conversion goals.
- Updated conversion goal change handling to reflect user input more accurately.
- Refactored button states to provide clearer feedback on actions.

This update streamlines the process of creating and managing dynamic conversion goals, enhancing overall usability.

Commit:
Merge branch 'feat/creating-conversion-goal-filter-in-dashboard' into chore/refactoring-columns-ma

github-actions[bot] committed:
Update UI snapshots for `chromium` (9)

Commit:
Pr's comments

Commit:
reverting change

Commit:
Merge remote-tracking branch 'origin/master' into feat/creating-conversion-goal-filter-in-dashboard

Commit:
Merge branch 'feat/creating-conversion-goal-filter-in-dashboard' of github.com:PostHog/posthog into feat/creating-conversion-goal-filter-in-dashboard

Commit:
Merge branch 'feat/creating-conversion-goal-filter-in-dashboard' into chore/refactoring-columns-ma

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

Commit:
update comment

Commit:
Merge branch 'chore/refactoring-columns-ma' of github.com:PostHog/posthog into chore/refactoring-columns-ma

Commit:
Merge remote-tracking branch 'origin/master' into chore/refactoring-columns-ma

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

Commit:
Merge branch 'feat/creating-conversion-goal-filter-in-dashboard' into chore/refactoring-columns-ma

Commit:
update comment

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

Commit:
fixing type

Commit:
Merge branch 'feat/creating-conversion-goal-filter-in-dashboard' of github.com:PostHog/posthog into feat/creating-conversion-goal-filter-in-dashboard

Commit:
Merge branch 'chore/refactoring-columns-ma' of github.com:PostHog/posthog into chore/refactoring-columns-ma

Commit:
merge conflicts

Commit:
Prs comments

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

Commit:
removing unused code

Commit:
fixing merge conflicts

Commit:
Fixing errors after refactor

Commit:
Merge remote-tracking branch 'origin/master' into chore/refactoring-columns-ma

Commit:
removing unnecessary action

Commit:
removing unnecessary var

Commit:
Fixing typeignore

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Commit:
merge conflicts

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

Commit:
Merge branch 'chore/refactoring-columns-ma' of github.com:PostHog/posthog into chore/refactoring-columns-ma

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

Commit:
Merge remote-tracking branch 'origin/master' into chore/refactoring-columns-ma

Commit:
Pr's comments

Commit:
Merge remote-tracking branch 'origin/master' into chore/refactoring-columns-ma

Commit:
undoing condition for change

Commit:
Merge remote-tracking branch 'origin/master' into chore/refactoring-columns-ma

Commit:
refactor(conversion_goal_processor): Update field access patterns to include 'events' prefix for consistency and clarity. Added comprehensive test suite for ConversionGoalProcessor covering various scenarios including node types, math types, and property filters.

Commit:
Removing team.id condition. Also handling chain event timestamp better

Commit:
Updating test to be production ready'

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (15)

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

github-actions[bot] committed:
Update UI snapshots for `chromium` (16)

github-actions[bot] committed:
Update UI snapshots for `chromium` (10)

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-tests-to-ma-query-conversion-goal

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Commit:
merge conflicts

Commit:
prettier

Commit:
Added xfail markers for tests related to unimplemented organic source logic and updated assertions to expect 'organic' as the source in multiple test cases. Adjusted campaign name expectations in edge case tests to reflect new logic.

Commit:
Add organic campaign and source constants; update attribution logic to default to organic

Commit:
removing unreachable code

Commit:
Fixing ruff.sh check

Commit:
uncommmit1

Commit:
first working plain sql

Commit:
first great approach

Commit:
only 2 tests left

Commit:
all tests passing

Commit:
Cleaning code

Commit:
Completly working with ast notation. Make sure you don't add all event goal

Commit:
Refactor ConversionGoalProcessor for clarity and efficiency. Removed unused logger, improved docstrings, and streamlined UTM expression handling. Updated query generation methods for better readability and maintainability. Adjusted test snapshots to reflect changes in query structure.

github-actions[bot] committed:
Update UI snapshots for `chromium` (14)

Commit:
bots comments

github-actions[bot] committed:
Update UI snapshots for `chromium` (13)

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

github-actions[bot] committed:
Update UI snapshots for `chromium` (10)

github-actions[bot] committed:
Update UI snapshots for `chromium` (4)

github-actions[bot] committed:
Update UI snapshots for `chromium` (11)

github-actions[bot] committed:
Update UI snapshots for `chromium` (14)

Commit:
conflicts merge

github-actions[bot] committed:
Update UI snapshots for `chromium` (13)

Commit:
Merge branch 'feat/adding-conversion-attribution-based-on-last-touching-point' of github.com:PostHog/posthog into feat/adding-conversion-attribution-based-on-last-touching-point

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-tests-to-ma-query-conversion-goal

github-actions[bot] committed:
Update UI snapshots for `chromium` (11)

github-actions[bot] committed:
Update UI snapshots for `chromium` (4)

Commit:
conflicts merge

Commit:
Merge branch 'feat/adding-tests-to-ma-query-conversion-goal' of github.com:PostHog/posthog into feat/adding-tests-to-ma-query-conversion-goal

Commit:
Merge branch 'feat/adding-tests-to-ma-query-conversion-goal' into feat/adding-conversion-attribution-based-on-last-touching-point

Commit:
Reverting csv copy on sql

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

github-actions[bot] committed:
Update UI snapshots for `chromium` (4)

github-actions[bot] committed:
Update UI snapshots for `chromium` (4)

Commit:
Adding one more test

Commit:
Refactor last_utm_timestamp calculation in ConversionGoalProcessor

Updated the logic for calculating last_utm_timestamp by replacing arrayElement and arraySort with arrayMax and arrayFilter for improved efficiency. Adjusted related test snapshots to reflect these changes.

Commit:
Add attribution mode and refactor conversion goal processor

Introduced  enum for handling attribution modes (LAST_TOUCH, FIRST_TOUCH). Updated  to use a max attribution window of 365 days and refactored related calculations to improve clarity. Added a new test for first-touch attribution with multiple touchpoints. Updated snapshots to reflect changes in query logic.

Commit:
Merge branch 'feat/adding-conversion-attribution-based-on-last-touching-point' of github.com:PostHog/posthog into feat/adding-conversion-attribution-based-on-last-touching-point

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

Commit:
python checks

Commit:
Merge branch 'feat/adding-conversion-attribution-based-on-last-touching-point' of github.com:PostHog/posthog into feat/adding-conversion-attribution-based-on-last-touching-point

Commit:
conflicts merge

robbie-c commented:
Looks like this PR has pulled in some changes from elsewhere, you might need to rebase both branches

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-tests-to-ma-query-conversion-goal

Commit:
Merge branch 'feat/adding-tests-to-ma-query-conversion-goal' into feat/adding-conversion-attribution-based-on-last-touching-point

Commit:
refactor(conversion_goal_processor): streamline attribute access and enhance action event filtering

- Replaced  calls with direct attribute access for , , , and uid=501(javierbahamondes) gid=20(staff) groups=20(staff),12(everyone),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),701(com.apple.sharepoint.group.1),33(_appstore),98(_lpadmin),100(_lpoperator),204(_developer),250(_analyticsusers),395(com.apple.access_ftp),398(com.apple.access_screensharing),399(com.apple.access_ssh),400(com.apple.access_remote_ae) in  for better readability and performance.
- Introduced  method to handle action events with specific date constraints.
- Updated tests to reflect changes in expected conversion counts and clarified assertions for better accuracy.
- Adjusted SQL generation in snapshots to ensure correct attribution logic for actions with multiple events.

Commit:
fix(conversion_goal): enhance validation and handling for DataWarehouseNode conversion goals

- Updated  to properly handle DataWarehouseNode fields, ensuring required fields are set and the 'event' field is removed to prevent validation errors.
- Added comprehensive validation checks in  for required fields like , , , and .
- Adjusted  to directly use validated ConversionGoalFilter objects when available, improving efficiency.

Commit:
conflicts merge

Commit:
fix(ConversionGoalDropdown): remove 'event' field to prevent validation errors in DataWarehouseNode

Commit:
removing unreachable code

lricoy commented on products/marketing_analytics/backend/hogql_queries/__snapshots__/test_conversion_goal_processor.ambr:
```diff
@@ -1,17 +1,236 @@
+              (
+  SELECT
+                  events.person_id,
+                  arrayFilter(x -> greater(x, 0), groupArray(if(equals(events.event, 'user signed up'), toUnixTimestamp(events.timestamp), 0))) AS conversion_timestamps,
+                  arrayFilter(x -> greater(x, 0), groupArray(if(equals(events.event, 'user signed up'), toFloat(1), 0))) AS conversion_math_values,
+                  arrayFilter(x -> notEmpty(toString(x)), groupArray(if(equals(events.event, 'user signed up'), toString(ifNull(events.properties.utm_campaign, '')), ''))) AS conversion_campaigns,
+                  arrayFilter(x -> notEmpty(toString(x)), groupArray(if(equals(events.event, 'user signed up'), toString(ifNull(events.properties.utm_source, '')), ''))) AS conversion_sources,
+                  arrayFilter(x -> greater(x, 0), groupArray(if(and(equals(events.event, '$pageview'), notEmpty(toString(ifNull(events.properties.utm_campaign, ''))), notEmpty(toString(ifNull(events.properties.utm_source, '')))), toUnixTimestamp(events.timestamp), 0))) AS utm_timestamps,
+                  arrayFilter(x -> notEmpty(x), groupArray(if(and(equals(events.event, '$pageview'), notEmpty(toString(ifNull(events.properties.utm_campaign, ''))), notEmpty(toString(ifNull(events.properties.utm_source, '')))), toString(ifNull(events.properties.utm_campaign, '')), ''))) AS utm_campaigns,
+                  arrayFilter(x -> notEmpty(x), groupArray(if(and(equals(events.event, '$pageview'), notEmpty(toString(ifNull(events.properties.utm_campaign, ''))), notEmpty(toString(ifNull(events.properties.utm_source, '')))), toString(ifNull(events.properties.utm_source, '')), ''))) AS utm_sources
```
If you can find a way to use a single array function here, I think it could speed up the query.

lricoy commented on products/marketing_analytics/backend/hogql_queries/conversion_goal_processor.py:
```diff
@@ -21,244 +20,954 @@
+        )
+
+    def _get_aggregation_expr(self) -> ast.Expr:
+        """Get aggregation expression based on math type"""
```
I think you can drop those `docstrings` that are redundant. I like the ones that explain a bit more, though.

lricoy commented on products/marketing_analytics/backend/hogql_queries/__snapshots__/test_conversion_goal_processor.ambr:
```diff
@@ -1,17 +1,236 @@
+              conversion_math_values[i] AS conversion_math_value,
+              conversion_campaigns[i] AS conversion_campaign,
+              conversion_sources[i] AS conversion_source,
+              arrayMax(arrayFilter(x -> and(lessOrEquals(x, conversion_timestamps[i]), greaterOrEquals(x, minus(conversion_timestamps[i], 15552000))), utm_timestamps)) AS last_utm_timestamp,
```
Did you try this with a window function approach like you did with the session duration ticket? I am not sure if it will work or be faster for your use case, but noting it so you can check.

```
lagInFrame(properties.utm_campaign) OVER (
          PARTITION BY person_id
          ORDER BY timestamp
          ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
      ) AS last_utm_campaign
```

lricoy reviewed (approved):
Clickhouse team can tell you more, but I think you can improve this one with some small changes (left some comments) but I guess you can do it on a follow up PR as well.

Approving to unblock, I will send you more stuff on slack :D
